# Classification models

In [7]:
import pandas as pd

In [8]:
df = pd.read_csv('data/df_0_oh.csv')

In [9]:
df.head()

Administrative  Administrative_Duration  Informational  \
0             0.0                      0.0            0.0   
1             0.0                      0.0            0.0   
2             0.0                     -1.0            0.0   
3             0.0                      0.0            0.0   
4             0.0                      0.0            0.0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0             1.0                      0.0   
1                     0.0             2.0                     64.0   
2                    -1.0             1.0                     -1.0   
3                     0.0             2.0                      0.0   
4                     0.0            10.0                    627.5   

   BounceRates  ExitRates  PageValues  SpecialDay  Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0      2               1.0   
1         0.00       0.10         0.0         0.0      2               2.0   
2         0.20       0.20         0.0         0.0      2               4.0   
3         0.05       0.14         0.0         0.0      2               3.0   
4         0.02       0.05         0.0         0.0      2               3.0   

   Browser  Region  TrafficType  Weekend  Revenue  ('New_Visitor',)  \
0        1       1            1        0        0               0.0   
1        2       1            2        0        0               0.0   
2        1       9            3        0        0               0.0   
3        2       2            4        0        0               0.0   
4        3       1            4        1        0               0.0   

   ('Other',)  ('Returning_Visitor',)  
0         0.0                     1.0  
1         0.0                     1.0  
2         0.0                     1.0  
3         0.0                     1.0  
4         0.0                     1.0

In [17]:
X = df.drop(columns='Revenue')
y = df['Revenue']

Divide in train and test

In [6]:
from sklearn.model_selection import train_test_split

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35)

In [35]:
from sklearn.metrics import confusion_matrix, classification_report

## Logistic Regression

In [1]:
from sklearn.linear_model import LogisticRegression

Using all the data:

In [24]:
logreg = LogisticRegression(max_iter = 10000)
logreg.fit(X, y)
logreg.score(X,y)

0.8452535760728218

Dividing train and test:

In [33]:
logreg = LogisticRegression(max_iter = 10000)
logreg.fit(X_train, y_train)
logreg.score(X_test,y_test)

0.8514046900394706

In [46]:
y_pred_logreg = logreg.predict(X_test)

In [48]:
print(classification_report(y_test, y_pred_logreg ))

              precision    recall  f1-score   support

           0       0.85      1.00      0.92      3662
           1       0.58      0.03      0.05       645

    accuracy                           0.85      4307
   macro avg       0.72      0.51      0.49      4307
weighted avg       0.81      0.85      0.79      4307



In [47]:
confusion_matrix(y_test, y_pred_logreg)

array([[3649,   13],
       [ 627,   18]])

True positives = 18
False positives = 13
False negatives = 627

## Linear SVC (Support Vector Classification)

In [2]:
from sklearn.svm import LinearSVC

In [29]:
linsvc = LinearSVC(max_iter=100000)
linsvc.fit(X,y)
linsvc.score(X,y)

/Users/lola/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


0.831111833550065

Using train and test data

In [42]:
linsvc = LogisticRegression(max_iter = 100000)
linsvc.fit(X_train, y_train)
linsvc.score(X_test,y_test)

0.8514046900394706

In [45]:
y_pred_linsvc = linsvc.predict(X_test)

In [44]:
confusion_matrix(y_test, y_pred_linsvc)

array([[3649,   13],
       [ 627,   18]])

## K-Nearest Neighbor (KNN)

In [30]:
from sklearn.neighbors import KNeighborsClassifier

In [32]:
knn = KNeighborsClassifier()
knn.fit(X,y)
knn.score(X,y)

0.8637841352405722

Using train and test data

In [49]:
knn = LogisticRegression()
knn.fit(X_train, y_train)
knn.score(X_test,y_test)

/Users/lola/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.8504759693522174

In [52]:
y_pred_knn = linsvc.predict(X_test)

In [53]:
confusion_matrix(y_test, y_pred_knn)

array([[3649,   13],
       [ 627,   18]])